gui_server_ecm_sync.md

```diff
@@ -0,0 +1,56 @@
+# Synchronization Between the GUI and Server ECM
+
+## Goal
+When creating, deleting, or modifying entities/components in the GUI _while paused_,
+GUI changes should be reflected in the server once simulation is resumed, or stepped while paused.
+
+## Approach
+If a user performs actions that modify/create/remove entities while simulation is paused, the ECM on the GUI side is updated directly to reflect these actions.
+The server's ECM does not reflect these user actions yet, but will be notified of the changes that took place once simulation is resumed or stepped while paused.
+When the server ECM is notified of these changes, it will be updated to have any changes made to the GUI ECM (unless the GUI and server are run in the same process, which will be discussed below).
+Once the server ECM is updated, simulation will be carried out as normal.
+
+The image below depicts how changes to the GUI's ECM are propogated to the server's ECM.
+The GUI and server interaction can be done through an event (green arrows), or through a service (red arrow).
```

Comment:
I was confused when I first read this, because events are intra-process and can't cross the GUI-server boundary. After looking closely at the image, I think what you mean is that the interaction can be initiated on the GUI side through an event, or skip the event and directly send a control message to the server, right?

I also believe that we've disabled the direct service calls for `ign-gazebo`, right? If that's true, we could just omit that to make the explanation more straightforward, eliminating the event vs service explanation.

Reply:
> I was confused when I first read this, because events are intra-process and can't cross the GUI-server boundary. After looking closely at the image, I think what you mean is that the interaction can be initiated on the GUI side through an event, or skip the event and directly send a control message to the server, right?

Yeah, that's a better way of putting it. Sorry for the confusion. I can re-word this so that it's clearer.

> I also believe that we've disabled the direct service calls for ign-gazebo, right? If that's true, we could just omit that to make the explanation more straightforward, eliminating the event vs service explanation.

From what I recall, we had a discussion about this and decided to support both the event and service so that we don't break behavior for downstream users who are depending on the event: https://github.com/ignitionrobotics/ign-gui/pull/306#discussion_r739588172

Unless other work on the model editor has disabled supporting the service communication, then we are supporting both the service and event approach.

Reply:
As a follow-up on the first part of my previous comment, I have reworded how the event works (GUI side only) in https://github.com/ignitionrobotics/design/pull/8/commits/dce6cae9cba7926636d5252dc53d37e636d89e25